src/shims/tls.rs

```diff
@@ -74,38 +79,62 @@ impl<'tcx> TlsData<'tcx> {
     pub fn load_tls(
         &self,
         key: TlsKey,
+        thread_id: ThreadId,
         cx: &impl HasDataLayout,
     ) -> InterpResult<'tcx, Scalar<Tag>> {
         match self.keys.get(&key) {
-            Some(&TlsEntry { data, .. }) => {
-                trace!("TLS key {} loaded: {:?}", key, data);
-                Ok(data.unwrap_or_else(|| Scalar::null_ptr(cx).into()))
+            Some(TlsEntry { data, .. }) => {
+                let value = data.get(&thread_id).cloned();
+                trace!("TLS key {} for thread {:?} loaded: {:?}", key, thread_id, value);
+                Ok(value.unwrap_or_else(|| Scalar::null_ptr(cx).into()))
             }
             None => throw_ub_format!("loading from a non-existing TLS key: {}", key),
         }
     }
 
-    pub fn store_tls(&mut self, key: TlsKey, new_data: Option<Scalar<Tag>>) -> InterpResult<'tcx> {
+    pub fn store_tls(
+        &mut self,
+         key: TlsKey, thread_id: ThreadId, new_data: Option<Scalar<Tag>>) -> InterpResult<'tcx> {
         match self.keys.get_mut(&key) {
             Some(TlsEntry { data, .. }) => {
-                trace!("TLS key {} stored: {:?}", key, new_data);
-                *data = new_data;
+                match new_data {
+                    Some(ptr) => {
```

Comment:
Why do you call this a `ptr`? It's any kind of scalar.

Also, I know we didn't do this before, but could you add an `assert!` that this is not NULL? Just to make sure the comment we have in `TlsEntry` is accurate.

Reply:
> Also, I know we didn't do this before, but could you add an assert! that this is not NULL? Just to make sure the comment we have in TlsEntry is accurate.

What is the idiomatic way to do that without the interpreter context? Or should I add it as an argument?

Reply:
Oh, we do not have `ecx` here, I guess that's why we did not have the assertion all along...

Yeah, nothing we can do here right now I think. I am wondering if using `Option` here was indeed a good idea. After this PR lands, I might try to change that and see how it looks.